File: model/mixing/tcn.py
"""TCN (Temporal Convolutional Network) Mixing Layer

Non-causal dilated depthwise-separable convolution으로 토큰 믹싱.
인코더용이므로 symmetric (non-causal) padding 사용.
O(T × kernel_size × n_dilations) 복잡도, 완전 병렬화.

CPU 최적: depthwise conv는 MKL-DNN이 AVX2/VNNI 최적화,
각 채널 독립으로 캐시 친화적.
"""
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import Tensor

from model.mixing.base import MixingLayer
from model.bitlinear import BitLinear


class TCNMixing(MixingLayer):
    """Non-causal dilated depthwise conv + pointwise BitLinear

    구조:
        [depthwise_conv(dilation=2^i) for i in range(n_dilations)]
        → 합산 → BitLinear pointwise projection

    Receptive field = kernel_size × sum(dilations)
    k=7, n_dilations=6: RF = 7 × 63 = 441 per layer
    """

    def __init__(self, cfg):
        super().__init__()
        d = cfg.d_model
        k = cfg.tcn_kernel_size
        n_dil = cfg.tcn_n_dilations

        # Depthwise dilated convolutions
        self.convs = nn.ModuleList()
        for i in range(n_dil):
            dilation = 2 ** i
            # symmetric padding for non-causal
            padding = (k - 1) * dilation // 2
            self.convs.append(
                nn.Conv1d(d, d, k, padding=padding, dilation=dilation, groups=d, bias=False)
            )

        # Pointwise projection (ternary)
        self.proj = BitLinear(d, d)

    def forward(self, x: Tensor, pad_mask: Tensor | None = None) -> Tensor:
        # x: (B, T, d) → (B, d, T) for Conv1d
        h = x.transpose(1, 2)

        # 다중 dilation 합산
        out = self.convs[0](h)
        for conv in self.convs[1:]:
            c = conv(h)
            # 패딩으로 인해 길이가 다를 수 있으므로 맞춤
            if c.size(2) != out.size(2):
                c = c[:, :, :out.size(2)]
            out = out + c

        out = F.relu(out)

        # (B, d, T) → (B, T, d)
        out = out.transpose(1, 2)

        # Pointwise projection
        out = self.proj(out)

        if pad_mask is not None:
            out = out * pad_mask.unsqueeze(-1).to(out.dtype)

        return out

File: inference_dense/src/mixing/tcn.rs
//! TCN mixing layer — Non-causal dilated depthwise conv + BitLinear pointwise
//!
//! depthwise_conv1d_avx2 C 커널 사용.
//! 본질적으로 양방향 (non-causal symmetric padding).

use std::collections::HashMap;
use anyhow::Result;

use crate::bmmq::TensorData;
use crate::common::*;
use super::MixingLayer;

/// 단일 dilation의 depthwise conv 가중치
struct DepthwiseConv {
    weight: Vec<f32>,  // (d_model, kernel_size) — row-major
    kernel_size: usize,
    dilation: usize,
    d_model: usize,
}

impl DepthwiseConv {
    fn forward(&self, x: &[f32], seq_len: usize, out: &mut [f32]) {
        unsafe {
            depthwise_conv1d_avx2(
                x.as_ptr(),
                self.weight.as_ptr(),
                std::ptr::null(),  // no bias
                out.as_mut_ptr(),
                seq_len as i32, self.d_model as i32,
                self.kernel_size as i32, self.dilation as i32,
            );
        }
    }
}

pub struct TCNMixing {
    convs: Vec<DepthwiseConv>,
    proj: BitLinear,
    d_model: usize,
}

impl TCNMixing {
    pub fn load_bmmq(
        tensors: &mut HashMap<String, TensorData>,
        prefix: &str, d_model: usize, kernel_size: usize, n_dilations: usize,
    ) -> Result<Self> {
        let mut convs = Vec::new();
        for i in 0..n_dilations {
            let key = format!("{}.convs.{}.weight", prefix, i);
            // Conv1d weight: (d_model, 1, kernel_size) in PyTorch → (d_model, kernel_size)
            let weight = bmmq_take_f32(tensors, &key)?;
            convs.push(DepthwiseConv {
                weight,
                kernel_size,
                dilation: 1 << i,
                d_model,
            });
        }
        let proj = BitLinear::load_bmmq(tensors, &format!("{}.proj", prefix))?;
        Ok(Self { convs, proj, d_model })
    }
}

impl MixingLayer for TCNMixing {
    fn forward_batch(
        &self, x: &[f32], seq_len: usize, d_model: usize,
        out: &mut [f32], bufs: &mut BatchBufs,
    ) {
        let total = seq_len * d_model;
        let mut acc = vec![0.0f32; total];
        let mut conv_out = vec![0.0f32; total];

        // 다중 dilation 합산
        for conv in &self.convs {
            conv.forward(x, seq_len, &mut conv_out);
            for i in 0..total {
                acc[i] += conv_out[i];
            }
        }

        // ReLU
        for v in acc.iter_mut() {
            *v = relu_scalar(*v);
        }

        // Pointwise BitLinear projection
        self.proj.forward_batch(&acc, seq_len, out, bufs);
    }

    fn forward_bidirectional(
        &self, x: &[f32], seq_len: usize, d_model: usize,
        out: &mut [f32], bufs: &mut BatchBufs,
        _rev_buf: &mut Vec<f32>, _fwd_buf: &mut Vec<f32>, _bwd_buf: &mut Vec<f32>,
    ) {
        // TCN은 non-causal → 본질적으로 양방향
        self.forward_batch(x, seq_len, d_model, out, bufs);
    }
}
